File: zero-downtime-migration/python/scripts/backfill_names.py
# scripts/backfill_names.py
# Backfills data into newly added columns without downtime
# This runs AFTER the schema expansion and while the app is live.

from database import SessionLocal
from models import User

# Create a database session
db = SessionLocal()

# Select only users that have NOT been backfilled yet
# This makes the script:
# - Safe to rerun
# - Incremental
# - Non-destructive

users = db.query(User).filter(User.first_name.is_(None)).all()

for user in users:
    # Skip users that do not have the old full_name field populated
    # (avoids corrupt or partial data)
    if not user.full_name:
        continue

    # Split the old full_name into first and last names
    # This is a simple heuristic for demo purposes
    parts = user.full_name.split(" ")
    user.first_name = parts[0]
    user.last_name = " ".join(parts[1:]) if len(parts) > 1 else ""

    # Mark the row as updated
    db.add(user)

# Commit all updates in a single transaction
# This avoids partial writes and keeps the operation atomic
db.commit()

# Confirmation log
print("✅ Backfill completed")

File: zero-downtime-migration/Typescript/src/scripts/backfill_names.ts
// ------------------------
// 1. Import dependencies
// ------------------------
import User from "../models/User"; // Sequelize User model
import sequelize from "../db";     // Sequelize instance

// ------------------------
// 2. Backfill function
// ------------------------
async function backfill() {
  // Fetch all users where firstName is still null
  const users = await User.findAll({
    where: { firstName: null }
  });

  // Loop through each user
  for (const user of users) {
    // Skip users that don't have a fullName
    if (!user.fullName) continue;

    // Split fullName into parts (assumes first space separates first and last names)
    const parts = user.fullName.split(" ");
    
    // Assign firstName to the first part
    user.firstName = parts[0];

    // Assign lastName to the rest of the name, join with space, default to empty string
    user.lastName = parts.slice(1).join(" ") || "";

    // Save changes to the database
    await user.save();
  }

  // Log confirmation after processing all users
  console.log("Backfill complete");

  // Exit the process after backfill is done
  process.exit(0);
}

// ------------------------
// 3. Sync database & run backfill
// ------------------------
sequelize.sync().then(backfill); // Ensures models are in sync before running backfill
